app/gui2/src/util/callTree.ts

```diff
@@ -210,78 +208,133 @@ export class ArgumentApplication {
     const notAppliedOriginally = new Set(
       noArgsCall?.notAppliedArguments ?? allPossiblePrefixArguments,
     )
-    const argumentsLeftToMatch = allPossiblePrefixArguments.filter(
-      (i) => !notAppliedSet.has(i) && notAppliedOriginally.has(i),
+    const argumentsLeftToMatch = allPossiblePrefixArguments.filter((i) =>
+      notAppliedOriginally.has(i),
     )
 
-    const prefixArgsToDisplay: Array<{
+    const resolvedArgs: Array<{
       appTree: Ast.Ast
       argument: ArgumentAst | ArgumentPlaceholder
     }> = []
 
-    function insertPlaceholdersUpto(index: number, appTree: Ast.Ast) {
-      let canInsertPositional = true
-      while (placeholdersToInsert[0] != null && placeholdersToInsert[0] < index) {
-        const argIndex = placeholdersToInsert.shift()
-        const argInfo = tryGetIndex(knownArguments, argIndex)
-        if (argIndex != null && argInfo != null) {
-          prefixArgsToDisplay.push({
-            appTree,
-            argument: ArgumentPlaceholder.WithRetrievedConfig(
-              callId,
-              argIndex,
-              argInfo,
-              ApplicationKind.Prefix,
-              !canInsertPositional,
-              widgetCfg,
-            ),
-          })
-
-          canInsertPositional = false
-        }
-      }
+    function nextArgumentName() {
+      return tryGetIndex(knownArguments, argumentsLeftToMatch[0])?.name
+    }
+
+    function takeNextArgument() {
+      const index = argumentsLeftToMatch.shift()
+      const info = tryGetIndex(knownArguments, index)
+      return index != null && info != null ? { index, info } : undefined
+    }
+
+    function takeNamedArgument(name: string) {
+      const takeIdx = argumentsLeftToMatch.findIndex(
+        (id) => tryGetIndex(knownArguments, id)?.name === name,
+      )
+      const index = argumentsLeftToMatch.splice(takeIdx, 1)[0]
+      const info = tryGetIndex(knownArguments, index)
+      return index != null && info != null ? { index, info } : undefined
     }
 
-    for (const realArg of interpreted.args) {
-      if (realArg.argName == null) {
-        const argIndex = argumentsLeftToMatch.shift()
-        if (argIndex != null) insertPlaceholdersUpto(argIndex, realArg.appTree.function)
-        prefixArgsToDisplay.push({
-          appTree: realArg.appTree,
+    function putBackArgument(index: number) {
+      argumentsLeftToMatch.unshift(index)
+    }
+
+    const lastPositionalArgIndex = findLastIndex(interpreted.args, (arg) => arg.argName == null)
+
+    let placeholderAlreadyInserted = false
+
+    let nextArgument: ReturnType<typeof takeNextArgument>
+
+    // Always insert a placeholder for the missing argument at the first position that is legal
+    // and don't invalidate further named arguments, treating the named arguments at correct
+    // position as if they were positional. foo a b c [d]
```

Comment:
> and don't invalidate further named arguments,

did you mean `positional arguments`? How could named be invalidated?

this `foo a b c [d]` is a sort of example?